x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/support/PointInTimeSearchHelper.java

```diff
@@ -0,0 +1,126 @@
+        executeAsyncWithOrigin(client, SECURITY_ORIGIN, OpenPointInTimeAction.INSTANCE, openPointInTimeRequest,
+            ActionListener.wrap(openPointInTimeResponse -> {
+                final String pitId = openPointInTimeResponse.getSearchContextId();
+                final OriginSettingClient securityOriginClient = new OriginSettingClient(client, SECURITY_ORIGIN);
+                getNextPage(new ArrayList<T>(), pitId, null, securityOriginClient, searchRequest, hitParser, listener);
+            }, listener::onFailure));
+    }
+
+    private static <T> void getNextPage(List<T> results, String pitId, Object[] sortValues,
+                                        Client client,
+                                        SearchRequest searchRequest,
+                                        Function<SearchHit, T> hitParser,
+                                        final ActionListener<Collection<T>> listener) {
+        assert client instanceof OriginSettingClient : "client should have a specific origin";
+        final PointInTimeBuilder pointInTimeBuilder = new PointInTimeBuilder(pitId).setKeepAlive(KEEP_ALIVE);
+        if (results.isEmpty()) {  // first page
+            assert sortValues == null : "search_after should be null for the first page";
+            logger.trace("getting first page with pit: [{}]", pitId);
+            searchRequest.source().pointInTimeBuilder(pointInTimeBuilder);
+        } else {
+            logger.trace("getting next page with pit: [{}] and sort values: [{}]",
+                pitId, Strings.arrayToCommaDelimitedString(sortValues));
+            searchRequest.source().searchAfter(sortValues).pointInTimeBuilder(pointInTimeBuilder);
+        }
+        client.threadPool().executor(ThreadPool.Names.SAME).execute(() -> {
+            client.execute(SearchAction.INSTANCE, searchRequest,
+                ActionListener.wrap(searchResponse -> {
+                    processSearchResponse(searchResponse, results, client, searchRequest, hitParser, listener);
+                }, listener::onFailure));
+        });
+    }
+
+    private static <T> void processSearchResponse(SearchResponse searchResponse, List<T> results,
+                                                  Client client, SearchRequest searchRequest,
+                                                  Function<SearchHit, T> hitParser,
+                                                  final ActionListener<Collection<T>> listener) {
+        assert client instanceof OriginSettingClient : "client should have a specific origin";
+        final SearchHit[] hits = searchResponse.getHits().getHits();
+        results.addAll(Arrays.stream(hits).map(hitParser).collect(Collectors.toUnmodifiableList()));
+        final String pitId = searchResponse.pointInTimeId();
+        // Not using totalHits as it must be interpreted with the relation, and tracking total hits is has cost
+        if (hits.length < searchRequest.source().size()) {
```

Comment:
I think we need a maximum cap here to avoid size of the result blowing up the node. 10,000 is a reasonable choice, the current [ScrollHelper](https://github.com/elastic/elasticsearch/blob/master/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/ScrollHelper.java#L71-L79) effectively does that as well though maybe unintentionally since the total hits value is bound by the relation, not the true total value.
The maximum cap should work for security need. But it does not solve the problem where there are indeed more than 10K results. To solve it, we need pagination for the user facing response which is a bigger problem and should be tackled separately.

Reply:
A 10k cap is fine for now.
We may hit some issues with that in Cloud security cluster, but simply loading everything into memory isn't the answer there.